selfdrive/loggerd/loggerd.cc

```diff
@@ -203,66 +303,46 @@ void encoder_thread(bool is_streaming, bool raw_clips, int cam_idx) {
+
+        pthread_mutex_lock(&s.rotate_lock);
+        pthread_mutex_unlock(&s.rotate_lock);
+
+        // wait if camera pkt id is older than stream
+        rotate_state->waitLogThread();
 
-      {
-        // all the rotation stuff
-        bool should_rotate = false;
-        std::unique_lock<std::mutex> lk(s.lock);
-        if (cam_idx == CAM_IDX_FCAM) { // TODO: should wait for three cameras on tici?
-          // wait if log camera is older on back camera
-          while ( extra.frame_id > s.last_frame_id //if the log camera is older, wait for it to catch up.
-                 && (extra.frame_id-s.last_frame_id) < 8 // but if its too old then there probably was a discontinuity (visiond restarted)
-                 && !do_exit) {
-            s.cv.wait(lk);
-          }
-          should_rotate = extra.frame_id > s.rotate_last_frame_id && encoder_segment < s.rotate_segment && s.rotate_seq_id == my_idx;
-        } else {
-          // front camera is best effort
-          should_rotate = encoder_segment < s.rotate_segment && s.rotate_seq_id == my_idx;
-        }
         if (do_exit) break;
 
         // rotate the encoder if the logger is on a newer segment
-        if (should_rotate) {
-          LOG("rotate encoder to %s", s.segment_path);
-
+        if (rotate_state->should_rotate) {
+          if (rotate_state->last_rotate_frame_id == 0) {
+            rotate_state->last_rotate_frame_id = extra.frame_id - 1;
+          }
+          while (s.rotate_seq_id != my_idx && !do_exit) { usleep(1000); }
+          LOGW("camera %d rotate encoder to %s.", cam_idx, s.segment_path);
           encoder_rotate(&encoder, s.segment_path, s.rotate_segment);
           s.rotate_seq_id = (my_idx + 1) % s.num_encoder;
-
           if (has_encoder_alt) {
             encoder_rotate(&encoder_alt, s.segment_path, s.rotate_segment);
           }
-
           if (raw_clips) {
             rawlogger->Rotate(s.segment_path, s.rotate_segment);
           }
-
           encoder_segment = s.rotate_segment;
           if (lh) {
             lh_close(lh);
           }
           lh = logger_get_handle(&s.logger);
-        }
 
-        if (encoder.rotating) {
           pthread_mutex_lock(&s.rotate_lock);
           s.should_close += 1;
           pthread_mutex_unlock(&s.rotate_lock);
 
-          while(s.should_close > 0 && s.should_close < s.num_encoder) {
-            // printf("%d waiting for others to reach close, %d/%d \n", my_idx, s.should_close, s.num_encoder);
-            s.cv.wait(lk);
-          }
+          while(s.should_close > 0 && s.should_close < s.num_encoder && !do_exit) { usleep(1000); }
 
           pthread_mutex_lock(&s.rotate_lock);
-          if (s.should_close == s.num_encoder) {
-            s.should_close = 1 - s.num_encoder;
-          } else {
-            s.should_close += 1;
-          }
+          s.should_close = s.should_close == s.num_encoder ? 1 - s.num_encoder : s.should_close + 1;
+
           encoder_close(&encoder);
```

Comment:
encoder_close needs time to flush out omx buffers.maybe we can consider putting the call before waiting s.should_close == s.num_encoder